Cpp tests

ernestmc opened the pull request:
Connects to ros2/ros2#168

dirk-thomas commented on rosidl_generator_cpp/test/test_array_generator.hpp:
```diff
@@ -0,0 +1,270 @@
+template<std::size_t SIZE>
+void test_array_bool(std::array<bool, SIZE> * dst_array)
+{
+  int i;
```
Please declare the variable in the smallest scope possible (https://github.com/ros2/ros2/wiki/Developer-Guide#programming-conventions). In this case within the for loop.

Same below.

ernestmc replied:
Ok.

ernestmc replied:
Fixed.

dirk-thomas commented on rosidl_generator_cpp/test/test_interfaces.cpp:
```diff
@@ -106,3 +113,186 @@ TEST(Test_rosidl_generator_traits, has_fixed_size) {
+  test_array_bool<10>(pattern_bool);
+  std::copy_n(pattern_bool->begin(), 10, message.bool_value.begin());
+  assert_array<bool, 10>(*pattern_bool, message.bool_value);
+  delete pattern_bool;
```
Instead of allocation the variable with `new` and later use `delete` you could just use a local variable.

ernestmc replied:
You're right.

ernestmc replied:
Fixed this.

dirk-thomas commented on rosidl_generator_cpp/test/test_interfaces.cpp:
```diff
@@ -106,3 +113,186 @@ TEST(Test_rosidl_generator_traits, has_fixed_size) {
+  for (i = 0; i < SIZE; i++) {
+    ASSERT_EQ(a[i], b[i]);
+  }
+}
```
Why is this function necessary? There should be a comparison operator for `std::array` which does that already.

ernestmc replied:
I will look at the matcher functions for gtest.

ernestmc replied:
ASSERT_EQ works for std::arrays too, didn't know that. Replaced all the checks.

dirk-thomas commented on rosidl_generator_cpp/test/test_array_generator.hpp:
```diff
@@ -0,0 +1,270 @@
+    }
+    (*dst_array)[SIZE - 1] = UINT64_MAX;
+  }
+}
```
Can you unify these numerous `test_array_T` functions in a single but generic template function?

ernestmc replied:
I was thinking of doing that but I'm not sure how to get the type limit constants for each case. Do you think I can do it using conditional compilation?

dirk-thomas replied:
Can you pass the type as a template argument and the boundaries as function arguments?

ernestmc replied:
Reduced the functions to two template functions: one for integer types and one for floats.

ernestmc commented:
This is ready for review.

Added tests for the following types:
* primitives_static
* primitives_bounded
* primitives_unbounded
* static_array_bounded
* static_array_unbounded
* static_array_static
* bounded_array_bounded
* bounded_array_unbounded
* bounded_array_static
* unbounded_array_bounded
* unbounded_array_unbounded
* unbounded_array_static

```
$ valgrind --leak-check=yes ./test_interfaces_cpp
==21088== Memcheck, a memory error detector
==21088== Copyright (C) 2002-2013, and GNU GPL'd, by Julian Seward et al.
==21088== Using Valgrind-3.10.1 and LibVEX; rerun with -h for copyright info
==21088== Command: ./test_interfaces_cpp
==21088== 
Running main() from gtest_main.cc
[==========] Running 13 tests from 2 test cases.
[----------] Global test environment set-up.
[----------] 1 test from Test_rosidl_generator_traits
[ RUN      ] Test_rosidl_generator_traits.has_fixed_size
[       OK ] Test_rosidl_generator_traits.has_fixed_size (10 ms)
[----------] 1 test from Test_rosidl_generator_traits (22 ms total)

[----------] 12 tests from Test_messages
[ RUN      ] Test_messages.primitives_static
[       OK ] Test_messages.primitives_static (22 ms)
[ RUN      ] Test_messages.primitives_bounded
[       OK ] Test_messages.primitives_bounded (124 ms)
[ RUN      ] Test_messages.primitives_unbounded
[       OK ] Test_messages.primitives_unbounded (259 ms)
[ RUN      ] Test_messages.static_array_bounded
[       OK ] Test_messages.static_array_bounded (25 ms)
[ RUN      ] Test_messages.static_array_unbounded
[       OK ] Test_messages.static_array_unbounded (38 ms)
[ RUN      ] Test_messages.static_array_static
[       OK ] Test_messages.static_array_static (24 ms)
[ RUN      ] Test_messages.bounded_array_bounded
[       OK ] Test_messages.bounded_array_bounded (24 ms)
[ RUN      ] Test_messages.bounded_array_unbounded
[       OK ] Test_messages.bounded_array_unbounded (37 ms)
[ RUN      ] Test_messages.bounded_array_static
[       OK ] Test_messages.bounded_array_static (24 ms)
[ RUN      ] Test_messages.unbounded_array_bounded
[       OK ] Test_messages.unbounded_array_bounded (35 ms)
[ RUN      ] Test_messages.unbounded_array_unbounded
[       OK ] Test_messages.unbounded_array_unbounded (51 ms)
[ RUN      ] Test_messages.unbounded_array_static
[       OK ] Test_messages.unbounded_array_static (37 ms)
[----------] 12 tests from Test_messages (703 ms total)

[----------] Global test environment tear-down
[==========] 13 tests from 2 test cases ran. (774 ms total)
[  PASSED  ] 13 tests.
==21088== 
==21088== HEAP SUMMARY:
==21088==     in use at exit: 0 bytes in 0 blocks
==21088==   total heap usage: 609 allocs, 609 frees, 76,164 bytes allocated
==21088== 
==21088== All heap blocks were freed -- no leaks are possible
==21088== 
==21088== For counts of detected and suppressed errors, rerun with: -v
==21088== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

jacquelinekay commented on rosidl_generator_cpp/test/test_array_generator.hpp:
```diff
@@ -0,0 +1,78 @@
+
+/**
+ * Helper function to generate a test pattern for boolean type.
+ * Alternating true (even index) and false (odd index) pattern.
```
If you're doing Doxygen-style comments, can you `\param` tags for each parameter

```
* \param[in] container (comment about parameter...)
* \param[in] size (comment about parameter...
```

ernestmc replied:
Added Doxygen comments.

jacquelinekay commented on rosidl_generator_cpp/test/test_array_generator.hpp:
```diff
@@ -0,0 +1,78 @@
+ * Alternating true (even index) and false (odd index) pattern.
+ */
+template<typename C>
+void test_vector_bool(C & container, int size)
```
Consider using an unsigned type (like `size_t`) instead of `int` for this argument and loop iterations.

jacquelinekay commented on rosidl_generator_cpp/test/test_array_generator.hpp:
```diff
@@ -0,0 +1,78 @@
+  std::default_random_engine rand_generator;
+  std::uniform_int_distribution<typename C::value_type> randnum(min, max);
+
+  if (size > 0) {
```
Another good reason to make `size` an unsigned type is to get rid of this check. The check currently doesn't seem to do anything when `size` is negative, either--is a negative size an invalid argument?

dirk-thomas replied:
This check is crucial even when `size` is of type `size_t` since a value `0` needs to be handled differently.

ernestmc replied:
Changed to use size_t.

jacquelinekay commented on rosidl_generator_cpp/test/test_array_generator.hpp:
```diff
@@ -0,0 +1,78 @@
+  std::uniform_int_distribution<typename C::value_type> randnum(min, max);
+
+  if (size > 0) {
+    container[0] = min;
```
Ah, I guess the above check would be important even when size is unsigned, but it could be changed to size != 0. Would you consider putting `container[0] = min` outside of the if block though?

ernestmc replied:
If size is zero then the vector won't have space to store any value that's why the assignment should be inside the if block. In that case we would be testing empty vectors.

jacquelinekay commented on rosidl_generator_cpp/test/test_array_generator.hpp:
```diff
@@ -0,0 +1,78 @@
+ * Mininum and maximum values for the type and random numbers in the middle.
+ */
+template<typename C>
+void test_vector_float_type(C & container, int size,
```
Consider renaming these functions to `generate_test_array` or something similar.

In fact... I think you could do some extra work with templates to make the `float` and `int` variations of this function have the same name (but create different specializations based on the type). Then you could templatize the part in `test_interfaces` where you call the functions the reduce some redundant code, similar what I recently tried to do in my `rmw_freertps` pull request: https://github.com/ros2/rmw_freertps/blob/megagen/rosidl_typesupport_freertps_cpp/test/test_serialization.cpp

ernestmc replied:
Maybe you can help me on this one. It would be nice to have the tests use one function name but I can't seem to figure out how to specialize the function. I have the container type C and can get the C::value_type. I was thinking of having something like:
test_vector<C, bool>
test_vector<C, integer>
test_vector<C, float>

jacquelinekay commented on rosidl_generator_cpp/test/test_array_generator.hpp:
```diff
@@ -0,0 +1,78 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef TEST_ARRAY_GENERATOR_HPP_
```
I believe the include guard should be something like:

```
#ifndef ROSIDL_GENERATOR_CPP__TEST__TEST_ARRAY_GENERATOR_HPP
```

(although cpplint doesn't complain about it for me)

dirk-thomas replied:
Since the header doesn't get installed it shouldn't be a problem if cpplint is fine with it.

ernestmc replied:
Ok, if it's not a problem I'll leave it like it is.

jacquelinekay commented on rosidl_generator_cpp/test/test_array_generator.hpp:
```diff
@@ -0,0 +1,78 @@
+  }
+}
+
+#endif
```
You should add an inline comment for the include guard down here, e.g. `#endif  // TEST_ARRAY_GENERATOR_HPP_`

ernestmc replied:
Done.

jacquelinekay commented:
Good stuff overall, especially the random pattern generation (I think the fancy word for that is "fuzz testing") and using the MAX and MIN values for datatypes. I'll probably incorporate that into my tests for freertps serialization on my next pass of that.

My major feedback about this PR is what I said about consolidating the code with templates. Feel free to email me if you want more direction on how to do that.

ernestmc commented:
@jacquelinekay thanks for your comments. I'll be addressing them soon.

jacquelinekay commented:
@ernestmc So sorry for leading you down the wrong path--I looked a bit closer at the code and realized I should have said for you to use macros, not templates.

Check out this commit, to branch `cpp_tests_macros`:

https://github.com/ros2/rosidl/commit/240b28e7c244887c07b5e4eeda958e6cd5abf15f

If you think you understand that example, go ahead and try simplifying the rest of this test case with macros. Otherwise let me know and I can answer your questions, or finish it in that branch.

jacquelinekay commented:
@dirk-thomas or someone else may have strong feelings about where those testing macros are defined--usually macros should go in a separate header file, or at the top of the file.

ernestmc commented:
Thanks @jacquelinekay ! I managed to templatize the test pattern generator functions using type traits, so now one function name works for all. Now I'll write the macros following your model. The code should be much nicer now.

ernestmc commented:
Updated and putting for review.
The code is much nicer and readable. Templatized the test pattern generation functions, added macros and functions to take away the repeated code of the tests.

jacquelinekay commented:
Looks great!

I started CI jobs:
http://ci.ros2.org/job/ci_linux/894/
http://ci.ros2.org/job/ci_osx/749
http://ci.ros2.org/job/ci_windows/975

If those come out ok, I think we should merge this.

jacquelinekay commented:
The cpplint style checker tool on Jenkins wants `test_interfaces.cpp` to include `vector`:

http://ci.ros2.org/job/ci_linux/894/testReport/rosidl_generator_cpp/cpplint/build_include_what_you_use__4____home_rosbuild_ci_scripts_ws_src_ros2_rosidl_rosidl_generator_cpp_test_test_interfaces_cpp_176_/

ernestmc commented:
Good advice hehe. How can I add this linter so it runs when I build with ament?

jacquelinekay commented:
I have 2 ways of doing this locally:

1. source the `install/setup.bash` made by ament, then the command `ament_cpplint` will be on your PATH. You can call it on a folder or an individual file.
2. Run:
<path to ament script>/ament test --skip-build --skip-install --only rosidl_generator_cpp
You can omit the first path if your install/setup.bash is sourced.
This will run the entire test suite on the package in your workspace. The test suite includes all the linters we use (cpplint, uncrustify, cppcheck, etc.)
Of course, you can use different flags, e.g. if you want to rebuild and reinstall, omit "skip-build" and "skip-install", and if you want to test the entire workspace, don't use "only".

ernestmc commented:
I use `$ ament test src/ros2/rosidl/rosidl_generator_cpp`
and it runs 7 different tests:

1. test_interfaces_cpp
1. copyright
1. cppcheck
1. lint_cmake
1. pep8
1. pyflakes
1. uncrustify

jacquelinekay commented:
Interesting, I wonder why cpplint didn't run locally for you.

Maybe cpplint was added to the set of linters for this package recently and you need to rebase against master?

@dirk-thomas would you like Ernesto to squash this? I think it's ready to merge.

jacquelinekay commented:
ah, my bad, I didn't look at the status of the Windows build.

http://ci.ros2.org/job/ci_windows/975/

Sometimes getting aggressive with templates will cause Visual Studios to fail.
```
17:51:49   C:\J\workspace\ci_windows\ws\src\ros2\rosidl\rosidl_generator_cpp\test\test_interfaces.cpp(145): error C2668: 'test_vector_fill': ambiguous call to overloaded function [C:\J\workspace\ci_windows\ws\build\rosidl_generator_cpp\test_interfaces_cpp.vcxproj]
17:51:49   C:\J\workspace\ci_windows\ws\src\ros2\rosidl\rosidl_generator_cpp\test\test_interfaces.cpp(186): error C2668: 'test_vector_fill': ambiguous call to overloaded function [C:\J\workspace\ci_windows\ws\build\rosidl_generator_cpp\test_interfaces_cpp.vcxproj]
```

I will investigate this since I have a Windows VM set up.

ernestmc commented:
hmmm I think that could be happening because a bool type is also an integral type and both templates could be getting enabled. I'll try adding an extra check (not sure if it's possible)

ernestmc commented:
@jacquelinekay could you test in Windows with my last fix?

jacquelinekay commented:
Yep, building it now.

jacquelinekay commented:
Hmm, still doesn't build.

esteve commented on rosidl_generator_cpp/test/test_array_generator.hpp:
```diff
@@ -0,0 +1,92 @@
+ */
+template<typename C>
+typename std::enable_if<std::is_integral<typename C::value_type>::value &&
+~std::is_same<typename C::value_type, bool>::value>::type
```
I think this should be like this:
```cpp
!std::is_same<typename C::value_type, bool>::value>::type

```

Also, I find this idiom slightly more readable, which we use in several places:

```cpp
template<typename C, typename std::enable_if<std::is_integral<typename C::value_type>::value &&
!std::is_same<typename C::value_type>::value>::type * = nullptr>
bool test_vector_fill(C & container, size_t size, typename C::value_type min, typename C::value_type max)

```

This way the template part doesn't pollute with the function signature.

esteve commented:
@ernestmc @jacquelinekay yeah, it's our VS's friendly way of telling us that it doesn't fully support expression SFINAE and so both versions of `test_fill_vector` are enabled.

esteve commented:
@ernestmc @jacquelinekay or it could be because of the bitwise not, but it'd be strange if it only affected VS.

ernestmc commented:
I don't have VS to test. If correcting the logic negation doesn't solve it I'll have to change the approach in those templates.

esteve commented:
@ernestmc @jacquelinekay I'm having a look at it and see if I can come up with a fix.

jacquelinekay commented:
Thanks!

esteve commented:
@ernestmc @jacquelinekay I think I've fixed the disambiguation bug in https://github.com/ros2/rosidl/tree/cpp_tests-esteve

Here's the diff https://github.com/ros2/rosidl/compare/cpp_tests...cpp_tests-esteve?expand=1

However it doesn't seem to compile because a `char` is being passed here:
https://github.com/ros2/rosidl/blob/cpp_tests-esteve/rosidl_generator_cpp/test/test_array_generator.hpp#L65

via this:

https://github.com/ros2/rosidl/blob/cpp_tests/rosidl_generator_cpp/test/test_interfaces.cpp#L156

Which goes against the C++11 spec, and (surprise!) VS is the only one that implements it correctly:

https://connect.microsoft.com/VisualStudio/feedback/details/843503/char-is-invalid-template-argument-for-uniform-int-distribution

More info here (see the types accepted in `IntType`):

http://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution

ernestmc commented:
Thanks a lot @esteve ! I'll have to write a special for char.

ernestmc commented:
Ok, there it goes. Hope VS likes it.

esteve commented:
@ernestmc you're welcome, let me know if there's anything else you want me to test on Windows.

esteve commented:
@ernestmc I've tried this on Windows and it now fails with `uint8_t`, I think after fixing that, the PR will be free from errors. Almost there!

ernestmc commented:
@esteve similar problem as before?

esteve commented:
@ernestmc yeah :-( According to http://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution, `uint8_t` is not one of the types accepted by `std::uniform_int_distribution`, only these:

> IntType 	- 	The result type generated by the generator. The effect is undefined if this is not one of `short`, `int`, `long`, `long long`, `unsigned short`, `unsigned int`, `unsigned long`, or `unsigned long long`.

ernestmc commented:
Yes, these are non standard types and it depends on how the implementation defines them. I'd bet int8_t and uint8_t are based on char types so I'll add the exception for those

dirk-thomas commented:
On `master` the command `ament test --only rosidl_generator_cpp` runs more linters for me. The same for the build farm: http://ci.ros2.org/job/ci_linux/894/testReport/rosidl_generator_cpp/

@ernestmc You might want to check why you don't have all linters locally.

ernestmc commented:
@dirk-thomas I had to add `<test_depend>ament_cpplint</test_depend>` to `package.xml`

ernestmc commented:
@esteve could you test my last changes on Windows?

dirk-thomas commented:
Which `package.xml` file are you referring to? This PR doesn't contain any changes and the manifest already contains all common linters: https://github.com/ros2/rosidl/blob/master/rosidl_generator_cpp/package.xml#L21

ernestmc commented:
@dirk-thomas I was referring to precisely that file. I added the changes locally to test. Although lint_common is present for some reason it doesn't invoke cpplint unless I explicitly add it.

jacquelinekay commented:
It's because that test dependency was added to the package.xml after you branched this branch off of master, and you need to either merge or rebase on the master branch.

esteve commented:
@ernestmc running it now, cross your fingers! :wink:

dirk-thomas commented:
`ament_lint_common` (https://github.com/ament/ament_lint/blob/master/ament_lint_common/package.xml) does already included `ament_cmake_cpplint`. Can you check the content of the file locally? Are you having the latest version of the `ament_lint` repo?

ernestmc commented:
@dirk-thomas I guess I don't have the latest version of the ament_lint repo because `ament_cpplint` and `ament_clang_format` are commented out.

esteve commented:
@ernestmc yay, it worked! Thank you so much for your patience going through all our feedback. The current PR looks good to me and ready to be merged. Could you squash the commits and trigger a CI, just to make sure?

ernestmc commented:
:smile: I sure will @esteve. Thank you guys!

ernestmc commented:
Squashed and triggered CI jobs:
* http://ci.ros2.org/job/ci_windows/986/
* http://ci.ros2.org/job/ci_osx/758/
* http://ci.ros2.org/job/ci_linux/906/

ernestmc commented:
I get a warning in Linux but seems to be in the gtest code: http://ci.ros2.org/job/ci_linux/906/warnings17Result/
Besides that it looks good to merge.

dirk-thomas commented on rosidl_generator_cpp/test/test_interfaces.cpp:
```diff
@@ -34,6 +42,9 @@
 #include "rosidl_generator_cpp/msg/unbounded_array_unbounded.hpp"
 
+#define TEST_VECTOR_SIZE 10
+#define TEST_VECTOR_SIZE_2 3
```
The constant name is not very helpful. Is this the vector size for nested messages?

ernestmc replied:
Yes, the first one is the size of the primitives array and the second one is the size of nested message array.

dirk-thomas replied:
What about calling it `TEST_NESTED_VECTOR_SIZE` then?

ernestmc replied:
I changed it to something more meaningful.

dirk-thomas commented:
The warning seems to point to an empty line `test_interfaces.cpp:120`: https://github.com/ros2/rosidl/pull/64/files#diff-8402e744cd022a18b9569d4f75323991R120 Can you please check what the real line resulting in this warning is? I guess there will be a way to use `gtest` differently to avoid the warning. Worst case we could always exclude the specific warning as a last resort.

ernestmc commented:
@dirk-thomas according to [this](https://github.com/google/googletest/issues/322):
```
This is a GCC bug.  gtest isn't trying to actually covert false to a pointer.  
It just uses a Template Meta Programming trick that's blessed by the C++ 
standard to detect NULL literals.
```
Maybe we can exclude that warning?

gerkey commented:
@ernestmc Go ahead and disable that warning, then we'll be ready to merge.

ernestmc commented:
Locally disabled warnings.

esteve commented:
:tada: :balloon: awesome! I've deleted the branch, we delete the branch of a PR after it's been merged, but sometimes we forget too :-)

ernestmc commented:
:ok_hand:
